server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java

```diff
@@ -0,0 +1,1192 @@
+      final InputRow row,
+      final Supplier<Committer> committerSupplier
+  ) throws IndexSizeExceededException, SegmentNotWritableException
+  {
+    if (!identifier.getDataSource().equals(schema.getDataSource())) {
+      throw new IAE(
+          "Expected dataSource[%s] but was asked to insert row for dataSource[%s]?!",
+          schema.getDataSource(),
+          identifier.getDataSource()
+      );
+    }
+
+    final Sink sink = getSink(identifier);
+    int sinkRetVal;
+
+    try {
+      sinkRetVal = sink.add(row);
+    }
+    catch (IndexSizeExceededException e) {
+      // Try one more time after swapping, then throw the exception out if it happens again.
+      persistAll(committerSupplier.get());
+      sinkRetVal = sink.add(row);
+    }
+
+    if (!sink.canAppendRow() || System.currentTimeMillis() > nextFlush) {
+      persistAll(committerSupplier.get());
+    }
+
+    if (sinkRetVal < 0) {
+      throw new SegmentNotWritableException("Attempt to add row to swapped-out sink for segment[%s].", identifier);
+    } else {
+      return sink.getNumRows();
+    }
+  }
+
+  @Override
+  public List<SegmentIdentifier> getSegments()
+  {
+    return ImmutableList.copyOf(sinks.keySet());
+  }
+
+  @Override
+  public int getRowCount(final SegmentIdentifier identifier)
+  {
+    final Sink sink = sinks.get(identifier);
+
+    if (sink == null) {
+      throw new ISE("No such sink: %s", identifier);
+    } else {
+      return sink.getNumRows();
+    }
+  }
+
+  private Sink getSink(final SegmentIdentifier identifier)
+  {
+    Sink retVal = sinks.get(identifier);
+
+    if (retVal == null) {
+      retVal = new Sink(
+          identifier.getInterval(),
+          schema,
+          identifier.getShardSpec(),
+          identifier.getVersion(),
+          tuningConfig.getMaxRowsInMemory()
+      );
+
+      try {
+        segmentAnnouncer.announceSegment(retVal.getSegment());
+      }
+      catch (IOException e) {
+        log.makeAlert(e, "Failed to announce new segment[%s]", schema.getDataSource())
+           .addData("interval", retVal.getInterval())
+           .emit();
+      }
+
+      sinks.put(identifier, retVal);
+      sinkTimeline.add(retVal.getInterval(), retVal.getVersion(), identifier.getShardSpec().createChunk(retVal));
+    }
+
+    return retVal;
+  }
+
+  @Override
+  public <T> QueryRunner<T> getQueryRunnerForIntervals(final Query<T> query, final Iterable<Interval> intervals)
+  {
+    if (conglomerate == null) {
+      throw new IllegalStateException("Don't query me, bro.");
+    }
+
+    final Iterable<SegmentDescriptor> specs = FunctionalIterable
+        .create(intervals)
+        .transformCat(
+            new Function<Interval, Iterable<TimelineObjectHolder<String, Sink>>>()
+            {
+              @Override
+              public Iterable<TimelineObjectHolder<String, Sink>> apply(final Interval interval)
+              {
+                return sinkTimeline.lookup(interval);
```

Comment:
anyway, it's not gonna be null. lookup returns an empty list if it doesn't find anything.